### PR TITLE
Implemented Content-Security-Policy-Report-Only for Browser Cache

### PR DESCRIPTION
**Pull request by jacobd91**
To test enable Browser Cache and navigate to the Browser Cache admin page. Once there you should see a section for Content Security Policy and under it and it's fields you should see a similar section for Content Security Policy Report Only. This second section is applied as separate header and should apply in conjunction with the CSP. Apply different values for each and review a page load's headers to see what policies were applied vs reported on

### CODE REVIEW

**Review by cssjoe — CHANGES_REQUESTED** [review]
The `report-to` directive is missing in the settings.
If CSPRO is enabled, then `report-to` should be a required URI, or the header should not be applied.

Warning: Though the report-to directive is intended to replace the deprecated `report-uri` directive, `report-to` isn't supported in most browsers yet. So for compatibility with current browsers, while also adding forward compatibility when browsers get report-to support, you can specify both `report-uri` and `report-to`.
In browsers that support report-to, the `report-uri` directive will be ignored.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
Directives missing from CSP and CSPRO:
`child-src`, `manifest-src`, `script-src-elem`, `script-src-attr`, `style-src-elem`, `style-src-attr`, `worker-src`
These can be added in another PR.

File: BrowserCache_Plugin.php [hunk 1]
```diff
@@ -467,4 +471,24 @@ function filter_wp_headers( $headers, $wp ) {
...
+	 */
+	public function admin_notices() {
+		if ( $this->_config->get_boolean( 'browsercache.security.cspro' ) && empty( $this->_config->get_string( 'browsercache.security.cspro.reporturi' ) ) && empty( $this->_config->get_string( 'browsercache.security.cspro.reportto' ) ) ) {
+			$message = '<p>' . sprintf(
+				// translators: 1 opening HTML a tag to Browser Cache CSP-Report-Only settings, 2 closing HTML a tag.
+				esc_html__(
+					'The Content Security Policy - Report Only requires the "report-uri" and/or "report-to" directives. Please define one or both of these directives %1$shere%2$s.',
+					'w3-total-cache'
+				),
+				'<a href="' . Util_Ui::admin_url( 'admin.php?page=w3tc_browsercache#browsercache__security__cspro' ) . '" target="_blank" alt="' . esc_attr__( 'Browser Cache Content-Security-Policy-Report-Only Settings', 'w3-total-cache' ) . '">',
```
Comment (cssjoe) [inline BrowserCache_Plugin.php:488]: Remove the "`target="_blank"`" attribute to remain on the same page/tab when clicked.

**Review by cssjoe — CHANGES_REQUESTED** [review]
Check the placeholders.  Most have "`Example: 'self' 'unsafe-inline' *.domain.com`".  Some expected values do not match the placeholder.  Check the docs -- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy

File: BrowserCache_Page_View_SectionSecurity.php [hunk 2]
```diff
@@ -15,7 +15,7 @@
 
 $c         = Dispatcher::config();
 $fp_values = $c->get_array( 'browsercache.security.fp.values' );
-
+error_log('config = ' . print_r($c,true));
```
[the remark this answers is not in the document]
Comment (jacobd91) [reply BrowserCache_Page_View_SectionSecurity.php:18]: Removed.

File: BrowserCache_Environment.php [hunk 3]
```diff
@@ -442,29 +445,104 @@ private function rules_cache_generate_apache( $config ) {
...
-					   ( !empty( $img ) ? "img-src $img; " : "" ).
-					   ( !empty( $media ) ? "media-src $media; " : "" ).
-					   ( !empty( $object ) ? "object-src $object; " : "" ).
-					   ( !empty( $plugin ) ? "plugin-types $plugin; " : "" ).
-					   ( !empty( $form ) ? "form-action $form; " : "" ).
-					   ( !empty( $frame_ancestors ) ? "frame-ancestors $frame_ancestors; " : "" ).
-					   ( !empty( $sandbox ) ? "sandbox " . trim( $sandbox ) . "; " : "" ).
-					   ( !empty( $default ) ? "default-src $default;" : "" ), "; " );
+				$dir = rtrim(
+					( ! empty( $base ) ? "base-uri $base; " : "" ) .
```
[the remark this answers is not in the document]
Comment (jacobd91) [reply BrowserCache_Environment.php:463]: Updated.

File: BrowserCache_Page_View_SectionSecurity.php [hunk 4]
```diff
@@ -582,6 +582,26 @@
 			</p>
 		</th>
 	</tr>
+	<tr>
+		<th>
+			<label for="browsercache_security_csp_reporturi"><?php Util_Ui::e_config_label( 'browsercache.security.csp.reporturi' ); ?></label>
+		</th>
+		<td>
+			<input id="browsercache_security_csp_reporturi" type="text" name="browsercache__security__csp__reporturi"
+				<?php Util_Ui::sealing_disabled( 'browsercache.' ); ?> value="<?php echo esc_attr( $this->_config->get_string( 'browsercache.security.csp.reporturi' ) ); ?>" size="50" placeholder="Example: /csp-violation-report-endpoint/" />
```
[the remark this answers is not in the document]
Comment (jacobd91) [reply BrowserCache_Page_View_SectionSecurity.php:591]: Updated.

File: BrowserCache_Page_View_SectionSecurity.php [hunk 5]
```diff
@@ -785,6 +875,324 @@
...
+			</p>
+		</th>
+	</tr>
+	<tr>
+		<th>
+			<label for="browsercache_security_cspro_reporturi"><?php Util_Ui::e_config_label( 'browsercache.security.cspro.reporturi' ); ?></label>
+		</th>
+		<td>
+			<input id="browsercache_security_cspro_reporturi" type="text" name="browsercache__security__cspro__reporturi"
+				<?php Util_Ui::sealing_disabled( 'browsercache.' ); ?> value="<?php echo esc_attr( $this->_config->get_string( 'browsercache.security.cspro.reporturi' ) ); ?>" size="50" placeholder="Example: /csp-violation-report-endpoint/" />
```
[the remark this answers is not in the document]
Comment (jacobd91) [reply BrowserCache_Page_View_SectionSecurity.php:909]: Updated.

File: BrowserCache_Plugin.php (hunk 1 above)
Comment (jacobd91) [reply BrowserCache_Plugin.php:488]: Removed.

File: BrowserCache_Page_View_SectionSecurity.php [hunk 6]
```diff
@@ -582,6 +582,26 @@
...
+			<div><i><?php esc_html_e( 'Instructs the user agent to report attempts to violate the Content Security Policy. These violation reports consist of JSON documents sent via an HTTP POST request to the specified URI.', 'w3-total-cache' ); ?></i></div>
+		</td>
+	</tr>
+	<tr>
+		<th>
+			<label for="browsercache_security_csp_reportto"><?php Util_Ui::e_config_label( 'browsercache.security.csp.reportto' ); ?></label>
+		</th>
+		<td>
+			<input id="browsercache_security_csp_reportto" type="text" name="browsercache__security__csp__reportto"
+				<?php Util_Ui::sealing_disabled( 'browsercache.' ); ?> value="<?php echo esc_attr( $this->_config->get_string( 'browsercache.security.csp.reportto' ) ); ?>" size="50" placeholder="Example: csp-endpoint" />
```
[the remark this answers is not in the document]
Comment (jacobd91) [reply BrowserCache_Page_View_SectionSecurity.php:601]: I left the placeholder as csp-endpoint and updated the description under the setting to read as follows:
```
Defines a reporting endpoint "group" to which violation reports should to be sent.

The referenced group should be defined in either the Report-To or Reporting-Endpoints HTTP headers. These will need to be manually defined either via htaccess or another method of modifying HTTP headers.
```

File: BrowserCache_Environment.php [hunk 7]
```diff
@@ -446,29 +449,104 @@ private function rules_cache_generate_apache( $config ) {
...
+				);
 
 				if ( !empty( $dir ) ) {
 					$rules .= "    Header set Content-Security-Policy \"$dir\"\n";
 				}
 			}
 
+			if ( $config->get_boolean( 'browsercache.security.cspro' ) && ( ! empty( $config->get_string( 'browsercache.security.cspro.reporturi' ) ) || ! empty( $config->get_string( 'browsercache.security.cspro.reportto' ) ) ) ) {
+				$base = trim( $config->get_string( 'browsercache.security.cspro.base' ) );
+				$reporturi = esc_url( $config->get_string( 'browsercache.security.cspro.reporturi' ) );
```
[the remark this answers is not in the document]
Comment (jacobd91) [reply BrowserCache_Environment.php]: Changed.

**Pull request merged**